src/hotspot/share/gc/shenandoah/shenandoahNMethod.cpp

```diff
@@ -126,13 +126,13 @@ void ShenandoahNMethod::heal_nmethod(nmethod* nm) {
   assert(data->lock()->owned_by_self(), "Must hold the lock");
 
   ShenandoahHeap* const heap = ShenandoahHeap::heap();
-  if (heap->is_concurrent_mark_in_progress()) {
-    ShenandoahKeepAliveClosure cl;
-    data->oops_do(&cl);
-  } else if (heap->is_concurrent_weak_root_in_progress() ||
-             heap->is_concurrent_strong_root_in_progress() ) {
+  if (heap->is_concurrent_weak_root_in_progress() ||
+      heap->is_concurrent_strong_root_in_progress()) {
     ShenandoahEvacOOMScope evac_scope;
     heal_nmethod_metadata(data);
+  } else if (heap->is_concurrent_mark_in_progress()) {
+    ShenandoahKeepAliveClosure cl;
+    data->oops_do(&cl);
```

Comment:
AFAICT, all that this change does is swap the condition ordering. Is this significant in any way?

Reply:
Yes, we did this so the evacuation phase would take precedence over concurrent marking of the old generation. That is, it is, when old marking is happening, it may also be the case that `is_concurrent_weak_root_in_progress` or `is_concurrent_strong_root_in_progress` and `is_concurrent_mark_in_progress` may all be true. However, the generational mode's old collection doesn't directly evacuate, so when roots are in progress, we want the evacuation case.